### ISSUE TITLE
Configuration Validation

### ISSUE THREAD

**Issue by quasiben** [dask/dask#5695]
I wanted to start a discussion on how to improve dask configuration through validation. 

A common problem users experience (with regards to configuration) is invalid and/or missing values.  This can often bubble up to the user in obscure errors and only after the application has started.  

A simple example:

```python
In [1]: import dask

In [2]: dask.config.get("logging")
Out[2]: {'distributed': 'info', 'distributed.client': 'warning', 'bokeh': 'critical'}

In [3]: dask.config.set({"logging.distributed": "inf"})
Out[3]: <dask.config.set at 0x7f0ddec57ac8>

In [4]: from distributed import Client
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-4-fa1827681742> in <module>
----> 1 from distributed import Client

/datasets/bzaitlen/GitRepos/distributed/distributed/__init__.py in <module>
----> 1 from . import config
      2 from dask.config import config
      3 from .actor import Actor, ActorFuture
      4 from .core import connect, rpc
      5 from .deploy import LocalCluster, Adaptive, SpecCluster, SSHCluster

/datasets/bzaitlen/GitRepos/distributed/distributed/config.py in <module>
    134
    135
--> 136 initialize_logging(dask.config.config)

/datasets/bzaitlen/GitRepos/distributed/distributed/config.py in initialize_logging(config)
    131             _initialize_logging_new_style(config)
    132         else:
--> 133             _initialize_logging_old_style(config)
    134
    135

/datasets/bzaitlen/GitRepos/distributed/distributed/config.py in _initialize_logging_old_style(config)
     91     for name, level in loggers.items():
     92         if isinstance(level, str):
---> 93             level = logging_names[level.upper()]
     94         logger = logging.getLogger(name)
     95         logger.setLevel(level)

KeyError: 'INF'
```

With configuration validation in place, the user would have received an error at the time of setting the value to `inf`, an invalid logging value.

I think it would be beneficial to do one, possibly two things:

1) Schema-tize configuration -- strongly type configuration values for a given key
2) build cli tool to validate configuration is correct (dask-config)

Some potentially libraries to consider in pursuit of validation:
- https://schematics.readthedocs.io/en/latest/
- https://traitlets.readthedocs.io/en/stable/
- https://docs.python-cerberus.org/en/stable/index.html
- https://marshmallow.readthedocs.io/en/stable/

**Comment by TomAugspurger** [comment]
Thanks for opening this Ben. Both of those sound valuable.

Not to bring up complications too quickly but dask itself doesn't have any external dependencies. We could consider vendoring one of those validation libraries if necessary.

**Comment by quasiben** [comment]
No, I think it's a good thing to bring up external dependencies now.  `Trailets`, for example, depends on `ipython_genutils`, `six`, and `funcsigs`.

**Comment by TomAugspurger** [comment]
An issue that would have possible been caught by this: https://github.com/dask/dask-ml/issues/708#issuecomment-663540969

**Comment by jsignell** [comment]
mashmallow has no external dependencies and I have found it pleasant to use in other projects.

**Comment by jsignell** [comment]
Oh wait. I forgot that #6069 uses jsonschema so I think we can just add helper for:

```python
import yaml
from jsonschema import validate
from dask.config import config

with open("/home/julia/dask/dask/dask-schema.yaml") as f:
    d = yaml.safe_load(f)
    
validate(instance=config, schema=d)
```

To see what the failure looks like you can do:

```python
config["temporary-directory"] = 2
validate(instance=config, schema=d)
```

**Comment by jsignell** [comment]
I opened a draft PR to push this idea a bit.

**Comment by djhoese** [comment]
I just had a realization for a (possibly hacky) way to improve import time when validation is not needed, but still allow for validation in the usual case. You/we could use the `__debug__` flag:

https://docs.python.org/3/library/constants.html#debug__

which would allow you to only do validation in the default execution case, but if users were concerned about import time they could run their code with `python -O my_script.py` and dask would then not validate the dask configuration.

Just wanted to throw this out there as an option for having the "best" of both worlds.
